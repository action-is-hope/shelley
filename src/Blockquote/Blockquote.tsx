import React from "react";
import type { ReactNode } from "react";
import type { TextVolume } from "../typings/shared-types";
import { Text } from "../Text";
import { st, classes } from "./blockquote.st.css";

export interface BlockquoteProps
  extends React.HTMLAttributes<HTMLQuoteElement> {
  children: ReactNode;
  /* Describe the quote with inline elements like cite and/or links to the source. **/
  desc?: ReactNode | string;
  /** Changes the volume of the description. **/
  descVol?: TextVolume;
  /** Variant type. **/
  variant?: "informal" | "academic";
}

function Blockquote(props: BlockquoteProps, ref: React.Ref<HTMLQuoteElement>) {
  const {
    children,
    className: classNameProp,
    desc,
    descVol = 2,
    variant,
    ...rest
  } = props;

  return (
    <blockquote
      ref={ref}
      className={st(classes.root, { variant }, classNameProp)}
      {...rest}
    >
      <div className={classes.content}>{children}</div>
      {desc && (
        <Text elementType="footer" vol={descVol} className={classes.desc}>
          {desc}
        </Text>
      )}
    </blockquote>
  );
}

/**
 * Blockquote... */
const _Blockquote = React.forwardRef(Blockquote);
export { _Blockquote as Blockquote };
