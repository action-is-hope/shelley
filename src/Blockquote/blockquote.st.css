@st-namespace "Blockquote";

@st-import Text from "../Text/text.st.css";

.root {
  -st-states:
    variant(enum(informal, academic));
}

.content {}

.desc {
  -st-extends: Text;
}
