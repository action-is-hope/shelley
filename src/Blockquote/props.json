{"./src/Blockquote/Blockquote.tsx":[{"description":"","methods":[],"displayName":"Blockquote","props":{"descVol":{"defaultValue":{"value":"2","computed":false},"required":false}}}]}