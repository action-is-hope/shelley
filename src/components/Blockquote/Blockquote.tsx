/** Blockquote.tsx */
import type React from "react";
import type { ReactNode, VFC } from "react";
import type { TextVolume } from "../types";
import Text from "../Text/Text";
/* = Style API. */
import { st, classes } from "./blockquote.st.css";

export interface BlockquoteProps
  extends React.HTMLAttributes<HTMLQuoteElement> {
  children: ReactNode;
  /*
   * Describe the quote with inline elements like cite and/or links to the source.
   **/
  desc?: ReactNode | string;
  /**
   * Changes the volume of the description.
   **/
  descVol?: TextVolume;
  /**
   * Variant type.
   **/
  variant?: "informal" | "academic";
}

export const Blockquote: VFC<BlockquoteProps> = ({
  children,
  className: classNameProp,
  desc,
  descVol = 2,
  variant,
  ...rest
}) => (
  <blockquote
    className={st(classes.root, { variant }, classNameProp)}
    {...rest}
  >
    <div className={classes.content}>{children}</div>
    {desc && (
      <Text as="footer" vol={descVol} className={classes.desc}>
        {desc}
      </Text>
    )}
  </blockquote>
);

export default Blockquote;
