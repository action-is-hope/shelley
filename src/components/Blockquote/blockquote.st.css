/** blockquote.st.css */
@namespace "Blockquote";

@st-import Text from "../../components/Text/text.st.css";

.root {
  -st-states:
    variant(enum(informal, academic));
}

.content {}

.desc {
  -st-extends: Text;
}
