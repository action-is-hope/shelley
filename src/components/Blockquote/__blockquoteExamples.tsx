/* __blockquoteExamples.tsx */
import React from "react";
// Example Component
import Blockquote from "./Blockquote";
// Supporting Components
import { P, H2 } from "../../components/Text/Text";
import PropsDemo from "../../components_site/PropsDemo/PropsDemo";
import CodeSample from "../../components_site/CodeSample/CodeSample";
import grid from "../../projects/default/css/grid.st.css";

export const meta = {
  name: "Blockquote"
};

export const QuickRef = () => (
  <CodeSample className={grid.mid}>{`import Blockquote from "shelley-ui";\n
<Blockquote
  cite="The Guardian, accepts react node so tect or link"
  >
  <P>These findings suggest a substantial impact of mechanized bots in amplifying denialist messages about climate change, including support for Trumpâ€™s withdrawal from the Paris agreement.</P>
</Blockquote>`}</CodeSample>
);

// Props Demo - Used in our docs and as a 'style test harness' - a fancy way of saying 'style checker'.
export const ComponentDemo = () => {
  const [labelDemoProps, setLabelDemoProps]: any = React.useState([
    {
      name: "variant",
      label: "variant",
      type: "number",
      min: 0,
      max: 6,
      value: 0
    },
    {
      name: "cite",
      label: "cite",
      type: "text",
      value: "Donella Meadows, environmental scientist, 1941-2001 "
    }
    // {
    //   name: "citeAttr",
    //   label: "citeAttr",
    //   type: "text",
    //   value: "https://en.wikipedia.org/wiki/Donella_Meadows"
    // }
  ]);

  const quotes = [
    `A system is a set of related components that work together in a 
    particular environment to perform whatever functions are required to 
    achieve the system's objective.`,

    `Technology and the limits to growth`,

    `Faith in technology as the ultimate solution to all problems
    can thus divert our attention from the most fundamental
    problem-the problem of growth in a wfinite system and prevent
    us from taking effective action to solve it.`
  ];
  {
    /* https://www.bbc.co.uk/sounds/play/w3csyx5b - we've been here before! */
  }
  return (
    <PropsDemo
      demoProps={labelDemoProps}
      setDemoProps={setLabelDemoProps}
      tsExtends="HTMLQuoteElement"
      renderExample={
        <CodeSample>{`import { Blockquote } from "@action-is-hope/shelley";\n\n<Blockquote ${
          labelDemoProps[0].value > 0
            ? `\n  variant={${labelDemoProps[0].value}}`
            : `\n  variant={ false }`
        } ${
          labelDemoProps[1].value ? `\n  cite="${labelDemoProps[1].value}"` : ``
        } \n>
${labelDemoProps[0].value == 0 ? `<P vol={4}>${quotes[0]}</P>` : ``} ${
          labelDemoProps[0].value == 1 ? `<P>${quotes[0]}</P>` : ``
        } ${
          labelDemoProps[0].value == 2
            ? `<>
      <H2 vol={4}>{quotes[1]}</H2>
      <P>${quotes[2]}</P>
    </>`
            : ``
        } ${labelDemoProps[0].value > 2 ? `<P>${quotes[0]}</P>` : ``}
</Blockquote>`}</CodeSample>
      }
    >
      <Blockquote
        variant={
          labelDemoProps[0].value === 0 ? false : labelDemoProps[0].value
        }
        cite={labelDemoProps[1].value}
        // citeAttr={labelDemoProps[2].value}
      >
        {/* Render a few deferent examples depending on the variant. */}
        {labelDemoProps[0].value == 0 && <P vol={4}>{quotes[0]}</P>}
        {labelDemoProps[0].value == 1 && <P>{quotes[0]}</P>}
        {labelDemoProps[0].value == 2 && (
          <>
            <H2 vol={4}>{quotes[1]}</H2>
            <P>{quotes[2]}</P>
          </>
        )}
        {labelDemoProps[0].value > 2 && <P>{quotes[0]}</P>}
      </Blockquote>
    </PropsDemo>
  );
};

export const ComponentHTML = () => (
  <CodeSample language="html">{`<blockquote class="blockquote--root">
  <div class="blockquote--content">
    <!-- children  - Text or whatever you like. -->
    <p class="text--root text--vol4">
      A system is a set of related components that work together...
    </p>
  </div>
  <footer class="blockquote--footer">
    <cite class="text--root text--vol3 blockquote--cite">
      <a href="#0">Donella Meadows, environmental scientist, 1941-2001</a>
    </cite>
  </footer>
</blockquote>`}</CodeSample>
);

export const ComponentCSS = () => (
  <CodeSample fixedHeight>{`/** blockquote.st.css */

/* Root */
Blockquote {}

/* Parts */
Blockquote::content {}
Blockquote::cite {}
Blockquote::footer {}

/* Variants */
Blockquote.variant1 {}
Blockquote.variant2 {}
Blockquote.variant3 {}
Blockquote.variant4 {}
Blockquote.variant5 {}
Blockquote.variant6 {}`}</CodeSample>
);
