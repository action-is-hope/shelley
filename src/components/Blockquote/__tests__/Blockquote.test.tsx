import Blockquote from "../Blockquote";
import renderer from "react-test-renderer";

const quoteText =
  "We but mirror the world. All the tendencies present in the outer world are to be found in the world of our body. If we could change ourselves, the tendencies in the world would also change. As a man changes his own nature, so does the attitude of the world change towards him. This is the divine mystery supreme. A wonderful thing it is and the source of our happiness. We need not wait to see what others do.";
const citeText = "Mahatma Gandhi";

describe("Blockquote", () => {
  it("renders as a basic blockquote with custom #className and data attr to test spread.", () => {
    const tree = renderer
      .create(
        <Blockquote
          data-testid="blockquote-data-testid"
          className="for-which-nobody-can-deny"
          desc={citeText}
        >
          {quoteText}
        </Blockquote>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders as a blockquote with description as a <cite> node at a larger volume.", () => {
    const tree = renderer
      .create(
        <Blockquote
          variant="academic"
          className="for-which-nobody-can-deny"
          desc={<cite>{citeText}</cite>}
          descVol={3}
        >
          {quoteText}
        </Blockquote>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
