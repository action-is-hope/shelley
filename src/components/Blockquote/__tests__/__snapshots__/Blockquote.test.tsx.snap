// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockquote renders as a basic blockquote with custom #className and data attr to test spread. 1`] = `
<blockquote
  className="Blockquote2773719531__root for-which-nobody-can-deny"
  data-testid="blockquote-data-testid"
>
  <div
    className="Blockquote2773719531__content"
  >
    We but mirror the world. All the tendencies present in the outer world are to be found in the world of our body. If we could change ourselves, the tendencies in the world would also change. As a man changes his own nature, so does the attitude of the world change towards him. This is the divine mystery supreme. A wonderful thing it is and the source of our happiness. We need not wait to see what others do.
  </div>
  <footer
    className="Text408270608__root Text408270608---vol-1-2 Blockquote2773719531__desc"
  >
    Mahatma Gandhi
  </footer>
</blockquote>
`;

exports[`Blockquote renders as a blockquote with description as a <cite> node at a larger volume. 1`] = `
<blockquote
  className="Blockquote2773719531__root Blockquote2773719531---variant-8-academic for-which-nobody-can-deny"
>
  <div
    className="Blockquote2773719531__content"
  >
    We but mirror the world. All the tendencies present in the outer world are to be found in the world of our body. If we could change ourselves, the tendencies in the world would also change. As a man changes his own nature, so does the attitude of the world change towards him. This is the divine mystery supreme. A wonderful thing it is and the source of our happiness. We need not wait to see what others do.
  </div>
  <footer
    className="Text408270608__root Text408270608---vol-1-3 Blockquote2773719531__desc"
  >
    <cite>
      Mahatma Gandhi
    </cite>
  </footer>
</blockquote>
`;
